### DISCUSSION TITLE
Newbie: explicit 3-way diff?

### DISCUSSION THREAD

**Discussion by akavel** [jj-vcs/jj#4573] [Q&A]
Hi!
For context, I'm a newbie to jj, and didn't yet use it much in practice; that said, I do have some working knowledge of git (but not deep internals, and not all its commands/flags/incantations either).
I'd like to ask: in jj, given some 3 branches A, B, C, I would like to build a "three-way diff", for a user to be able to fully resolve/accept each difference/chunk by hand, for example with the kdiff3 GUI tool (or a builtin one in jj if present). The thing is,

 - I want to make sure the user sees ***every*** difference, not just the conflicting ones - also the "non-conflicting" ones - and must manually approve/reject/choose resolution to every change;
 - this 3-way diff tool should always get called, unless those 3 commits compared are exactly identical.

I'm fine to call this tool on separate workspaces if that's the only way; that said, in such case I'd still be grateful for help how to do this and how to then build a final "merge commit" if the user accepted, that would include the specific final resolutions they selected for each "discrepancy area".

I'd be also super grateful if you could help me understand how I could do it via jujutsu's Rust API as well - if all goes well, I'd actually like to try and use this conflict resolution "embedded" in a utility app I'm trying to write in Rust.

TIA!

**Comment by martinvonz** [comment]
I think you're looking for the documentation here: https://martinvonz.github.io/jj/latest/config/#3-way-merge-tools-for-conflict-resolution. Check that out first. Perhaps your own merge tool can work with that file-based API.

**Reply by arxanas** [reply]
If writing your own merge tool, `scm-diff-editor` could be a good place to start, as it accepts the files that jj provides, then does its own three-way merge; follow where `base` is used here:

https://github.com/arxanas/scm-record/blob/8c2d4f36f933ed02acd0a147289925b224deb948/scm-diff-editor/src/lib.rs#L56-L65

If suitable, you could also contribute directly to `scm-record`/`scm-diff-editor`.

**Comment by ilyagr** [comment]
It's not immediately clear to me what you need help with.

@martinvonz will correct me if I'm wrong, but AFAIK `jj resolve` should pass the exact original files to `kdiff3`, with one exception: if two files are identical in the base of the conflict and one of the sides, that conflict is auto-resolved. Ideally, `jj` would have a way to customize this, but it's a tricky question of how the UI for this would work and would require some changes to how `jj` handles conflicts.

-----------

Another issue you may be having is that we configured `kdiff3` to auto-resolve conflicts by default with its `--auto` option. You can undo this by configuring:

```toml
[merge-tools.kdiff3]
merge-args = ["$base", "$left", "$right", "-o", "$output"]
```

or you can create an extra merge tool `kdiff3-noauto` following the docs Martin linked. See also `kdiff3 --help` for more options. By contrast, the default config is:

```console
$ jj config list --include-defaults merge-tools.kdiff3
merge-tools.kdiff3.merge-args = ["$base", "$left", "$right", "-o", "$output", "--auto"]
merge-tools.kdiff3.edit-args = ["--merge", "--cs", "CreateBakFiles=0", "--cs", "SkipDirStatus=1", "$left", "$right", "$output"]
```

--------------

There's also the issue that `jj resolve` currently works one file at a time. You can still run it in a loop or combine it with `jj resolve --list`. Or, you could implement the ability for it to work on entire dirs. The main reason this is not implemented is that few merge tools can handle it, but `kdiff3` is one of the tools that can (though I find the interface it uses for it un-intuitive).

BTW, the paid [Beyond Compare](https://www.scootersoftware.com/) has similar abilities to `kdiff3` and might have an easier to use interface for merging three dirs, though YMMV. They do have a generous free trial (30 days, with only days that you actively use the program counted).

-----------

Finally, if all else doesn't work (e.g. if the the auto-resolution of whole-file conflicts makes `jj resolve` a non-starter), you can do it manually with workspaces, as you suggested. 

If there is no relationship between `A`, `B`, and `C` and you want to edit one (or more) of those three commits, you can manually create three workspaces with `jj workspace add dir_A -r A; jj workspace add dir_B -r B; jj workspace add dir_C -r C` and then run 
either `meld` or https://github.com/ilyagr/diffedit3 on these three dirs (meld will allow you edit all three sides, `diffedit3` will only let you edit one of the sides). I don't think particular configuration is doable with `kdiff3`[^1], but I'm not sure this is what you wanted either.

My best guess is that you wanted to treat `B` as the base of the merge and then output the merge to a fourth commit. Then, you should do `jj new some_parents; jj workspace add dir_output` in addition to the above. Then, do `kdiff3 dir_B dir_A dir_C -o dir_output`. This is again the interface I mentioned as unintuitive above.

[^1]: IIRC, `kdiff3 dir_B dir_A dir_C` results in a read-only interface. Your best bet with `kdiff3` and 3 dirs would be to try `kdiff3 dir_B dir_A dir_C --merge`, I found it quite confusing and am not immediately sure what it does or what it's good for, but perhaps it will be helpful to you. The four-dir version I explain at the end is probably more sensible.

**Reply by akavel** [reply]
Thanks a lot! Auto-resolution of anything is indeed a non-starter for me, so I guess I will need to look at the workspaces in such case. Huge thanks for all the details and the examples! I don't think I realized kdiff3 actually has difficulties when working with whole directory trees in case of a 3-way merge; I admit it was quite long ago that I used kdiff3 for some complex cases, so my memory on that is definitely super hazy.

As for what is my specific use-case: I'll try to explain it, but please forgive me if it's messy, as I don't find it super easy to do, even more so in a simple way. Basically, I'm trying to write a "configuration management" tool (think more or less Ansible, or Terraform, or more distantly Nix). Now, this tool is stateful, storing the expected state of a managed machine as a files tree in a git repository. 

Storing the configuration state in a git repository is intended to allow good versioning and - here comes our use-case - **precise control over conflict resolution**. The app is intended to present the user with a "3-way diff" of the following trees:

- A (base) - last known historical "consistent" state of the managed machine;
- B - the state currently observed on the managed machine (could have diverged from A);
- C - the state the user said they want the machine to have (it is rendered from a script they provide, specifically written in [Nickel](https://github.com/tweag/nickel), which is an attempt at a modernized Nix language).

Notably, each of A, B, and C, can be different. And I want to present the user with a 3-way diff view of those three trees, giving them an option of what they want to "commit" and have the tool apply to the managed machine. (Thus, indeed, building a fourth tree D, to be added as a commit on top of A.)

I don't want any auto-resolution, because I want the user to have full control and visibility into each of the three trees A, B, and C. And I want to always present this UI before a commit, because notably even if A==B, I still want to make sure each of the changes in C is scrutinized, as they are generated from a script, so I want the user to be able to see what the script ended up emitting, and compare it with the existing state. OTOH, if B==C, I also want the user to clearly see that, and also see whether and where (if anywhere) A!=B.

Is the explanation above somewhat understandable, or should I try to clarify some parts?

**Reply by ilyagr** [reply]
This is pretty clear. I just have a bunch of inconclusive thoughts and comments.

Repeating myself, in theory `jj` should eventually support this use-case, but in practice it doesn't quite. I personally probably won't work on this in the near future, given the current state of available merge tools.

A bit of a warning: while in theory it's tempting to try to present all the information to the user in as raw a fashion as possible, in practice this often becomes an overwhelming amount of information. IMO, the most important job of the merge tools is to focus the user's attention on the most important small subset of the information available. What I *am* currently slowly working on is experimenting with interfaces that are simpler at the cost of hiding more information.

With you use-case, I don't know how confusing presenting all that info would be. The only way I know to find out is to build something and then to try using it for a bit.

Kdiff3 does share a similar philosophy of showing a lot of information. It definitely supports your usecase, though as I said I don't like using it for this purpose; this is also something you have to try. I am not aware of a free tool that's better than kdiff3 for this (or even try to support this usecase). Paid tools like Beyond Compare and probably Araxis also support it (you'd have to try them to see if they are better).

Personally, when not experimenting, I currently do all of my merging either using a text editor with the conflict markers or by using KDiff3 one-file-at-a-time, or by using 4-pane vimdiff setup one file-at-a-time (as set up by default as a merge tool for `jj`). I use vimdiff less and less as I've become more comfortable with `jj`'s conflict markers, but it was a valuable learning tool and is what I would use if I couldn't use kdiff3 for some reason.

BTW, vimdiff *could* be modified to create a 4-pane view for many files at a time (in different tabs), but I'm not aware of anybody having written the vimscript to do it. The closest I know of is https://github.com/balki/vimtabdiff, but it only supports comparisons of two sides. The problem, of course, is that only people comfortable with Vim can use that.

**Reply by ilyagr** [reply]
Also, at this point I'm not sure you are using any of jj's features. You can use `git worktree` just as well as `jj workspace`. Of course, jj can be useful for other things your to might do.

**Reply by arxanas** [reply]
@akavel Is this the same use-case/scenario as setting the difftool/mergetool in `.gitattributes`?

**Reply by akavel** [reply]
> Also, at this point I'm not sure you are using any of jj's features. You can use `git worktree` just as well as `jj workspace`. Of course, jj can be useful for other things your to might do.

Part of my reason for interest here was/is to maybe embed jj in my tool (which is also written in Rust) as a library. As I wrote in the initial question above:

> I'd be also super grateful if you could help me understand how I could do it via jujutsu's Rust API as well - if all goes well, I'd actually like to try and use this conflict resolution "embedded" in a utility app I'm trying to write in Rust.

**Reply by akavel** [reply]
> @akavel Is this the same use-case/scenario as setting the difftool/mergetool in `.gitattributes`?

@arxanas I'm afraid I don't understand what you mean by that. I don't know much about `.gitattributes`.

**Reply by ilyagr** [reply]
I haven't tried to use the jj API myself, but we provide some examples in the examples directory. There are people on discord who have done it. It'd be probably be simplest to see how jj cli's code for similar functionality works. 

Launching merge tools is currently part of the cli rather than the library, but it sounds like you won't actually use it.

If there's functionality in the cli you actually need, you'd have to first send a PR to move it into the library so you can use it via API. Or you could use the CLI, probably as an external process.

**Reply by arxanas** [reply]
@akavel gitattributes/custom difftools like this https://git-scm.com/docs/gitattributes#_defining_an_external_diff_driver
